content/blog/2024-08-29-valkey-memory-efficiency-8-0.md

```diff
@@ -0,0 +1,148 @@
+
+![Figure 3 Dictionary in cluster mode (Valkey 7.2) with multiple key value pair](/assets/media/pictures/dictionary_in_cluster_mode_7.2.png)
+
+## Improvements
+
+### Optimization 1 - [Dictionary per slot](https://github.com/redis/redis/pull/11695)
+
+The first optimization is we have introduced a dictionary per slot (16,384 of them in total), where each dictionary stores data for a given slot. With this simplification, the cost of maintaining additional metadata for the mapping of slot to key is no longer required in Valkey 8. To iterate over all the keys in a given slot, the engine simply finds out the dictionary for a given slot and traverse all the entries in it. This reduces the memory usage per dictionary entry by 16 bytes with a small memory overhead around 1 MB per node. As cluster mode is generally used for storing large amount of keys, avoiding the additional overhead per key allows users to store more number of keys in the same amount of memory.
+
+![Figure 4 Dictionary in cluster mode (Valkey 8.0) with multiple key value pair](/assets/media/pictures/dictionary_in_ cluster_mode_8.0.png)
+
+Few of the interesting challenges that comes up with the above improvements are supporting existing use cases which were optimized with a single dictionary for the entire keyspace. The usecases are:
+
+* [Iterating the entire keyspace](https://valkey.io/commands/scan/) - Command like `SCAN` to iterate over the entire keyspace.
+* [Random key for eviction](https://valkey.io/topics/lru-cache/) - The server does random sampling of the keyspace to find ideal candidate for eviction.
+* [Finding a random key](https://valkey.io/commands/randomkey/) - Command like `RANDOMKEY` retrieves a random key from the database.
+
+
+It is quite possible that some slots contains more keys than others and some could be also empty which could lead the engine to traverse through the all the empty dictionary for a full scan i.e. [O(N)](https://en.wikipedia.org/wiki/Big_O_notation) operation if approached naively. When picking a random key for eviction or running `RANDOMKEY` operation, the server needs to pick a non-empty slot in a performant manner. These cases require a data structure which provides the following functionality:
```

Comment:
`could lead the engine to traverse through the all the empty dictionary` -> `could lead the engine to traverse through the all empty dictionary` (I think that's what you meant).

Reply:
I meant `could lead the engine to traverse through all the empty dictionary`